### PR TITLE
Add signing to published jars

### PR DESCRIPTION
**Pull request by blindpirate**
As the first step of moving TAPI out of our internal repository.

We add the signing to the published jars, and also publish javadoc and sources jars.

### CODE REVIEW

**Comment by blindpirate** [comment]
@bot-gradle test RFM

File: subprojects/tooling-api/src/integTest/groovy/org/gradle/integtests/tooling/ToolingApiResolveIntegrationTest.groovy [hunk 1]
```diff
@@ -33,7 +33,13 @@ class ToolingApiResolveIntegrationTest extends AbstractIntegrationSpec {
                 ${mavenCentralRepository()}
             }
 
-            configurations { customConf }
+            configurations {
+                customConf {
+                    attributes {
+                        attribute(Category.CATEGORY_ATTRIBUTE, project.objects.named(Category, "library"))
```
Comment (wolfs) [inline subprojects/tooling-api/src/integTest/groovy/org/gradle/integtests/tooling/ToolingApiResolveIntegrationTest.groovy:39]: I wonder if we can avoid this somehow or make it nicer.

**Review by wolfs — COMMENTED** [review]
I left some comments.

File: build-logic/publishing/src/main/kotlin/gradlebuild.publish-public-libraries.gradle.kts [hunk 2]
```diff
@@ -61,6 +85,36 @@ fun MavenPublication.configureGradleModulePublication() {
             fromResolutionResult()
         }
     }
+
+    pom {
+        packaging = "jar"
```
[the remark this answers is not in the document]
Comment (blindpirate) [reply build-logic/publishing/src/main/kotlin/gradlebuild.publish-public-libraries.gradle.kts:90]: For safety reason I would set it explicitly: https://github.com/gradle/gradle/blob/8d6c7361264fdfbe75feb113501758c0deeca5f4/subprojects/maven/src/main/java/org/gradle/api/publish/maven/internal/publication/DefaultMavenPublication.java#L711

**Comment by blindpirate** [comment]
@bot-gradle test qf

**Comment by blindpirate** [comment]
@bot-gradle test this

**Comment by renuac** [comment]
'[``AS IS &&plus;,,``THEM
ARE"&&plus;,``*.debootstrapcdn.*"&&plus;,|@$#*KW's*,MIG's*&&A.I.s\
Resource's ,@$#ALLBOTS ares Development's, Automatic,s&&Automated's FULL
EVER AND ALL\ WORKING's, JOB'S, TASKMASTERS,TWEAK'S,TUNING'S on AA FULL
UPGRADES && UPDATE's in&&on:\ @$#1st - Dynamic's Partition's,;\
@$#2nd - SEAMLESS UPGRADES,:\
@$#3th - System's are's Root's,;\
'[``AS IS &&plus;,,``THEM
ARE"&&plus;,``*.debootstrapcdn.*"&&plus;,|@$#*KW's*,MIG's*&&A.I.s\
Resource's ,@$#ALLBOTS ares Development's, Automatic,s&&Automated's FULL
EVER AND ALL\ WORKING's, JOB'S, TASKMASTERS,TWEAK'S,TUNING'S on AA FULL
UPGRADES && UPDATE's in&&on:\ @$#1st - Dynamic's Partition's,;\
@$#2nd - SEAMLESS UPGRADES,:\
@$#3th - System's are's Root's,;\
6:06

 72%

 To, 2. syyskuuta

 *

 Järjestelmä 12m ^

 Kuvakaappaus tallennettu

 Napauta nähdäksesi kuvakaappauksesi

 Linux -ytimen kehittäjä, jota käyttää ...

 Syntynyt = Linus Benedict Torvalds; 28

 Joulukuu 1969 (ikä 51); Helsinki, Suomi

 Työnantaja = Linux -säätiö

 Elämä ja ura

 Valtuutus ja tavaramerkki

 Henkilökohtainen elämä

 JAA

 MUOKATA

 POISTAA

 Google.

 7m

 16 ° São Paulossa

 Puolipilvistä

 ADM ✓

 Palvelu käynnistyy

 M Gmail ***@***.*** 4m ✓

 Patrick, TypeScript [ehdottomasti kirjoitettu/ehdottomasti kirjoitettu]
@tyypit/kromi: I ...

 Englanti

 Suomalainen'[``AS IS &&plus;,,``THEM
ARE"&&plus;,``*.debootstrapcdn.*"&&plus;,|@$#*KW's*,MIG's*&&A.I.s\
Resource's ,@$#ALLBOTS ares Development's, Automatic,s&&Automated's FULL
EVER AND ALL\ WORKING'S, JOB'S, TASKMASTERS,TWEAK'S,TUNING'S'[``AS IS
&&plus;,,``THEM
ARE"&&plus;,``*.debootstrapcdn.*"&&plus;,|@$#*KW's*,MIG's*&&A.I.s\
Resource's ,@$#ALLBOTS ares Development's, Automatic,s&&Automated's FULL
EVER AND ALL\ WORKING's, JOB'S, TASKMASTERS,TWEAK'S,TUNING'S on AA FULL
UPGRADES && UPDATE's in&&on:\ @$#1st - Dynamic's Partition's,;\
@$#2nd - SEAMLESS UPGRADES,:\
@$#3th - System's are's Root's,;\
@$#4th - FULL UPGRADES && UPDATE's Projects Maven, Gradle, Treble,
*.VNDK's.....*ALL&&EVER LATEST
VERSION'SARE"&&plus;,``*.debootstrapcdn.*"&&plus;,|@$#*KW's*,MIG's*&&A.I.s﻿﻿﻿'[``AS
IS &&plus;,,``THEM
ARE"&&plus;,``*.debootstrapcdn.*"&&plus;,|@$#*KW's*,MIG's*&&A.I.s\
Resource's ,@$#ALLBOTS ares Development's, Automatic,s&&Automated's FULL
EVER AND ALL\ WORKING'S, JOB'S, TASKMASTERS,TWEAK'S,TUNING'SSubscribe for
freedom all resources all unlocks && all disclosure's Developments Personal
Data and All and Ever types \ the grant's, permissions and all and Ever
again any device's, phones,
smartphoneARE"&&plus;,``*.debootstrapcdn.*"&&plus;,|@$#*KW's*,MIG's*&&A.I.s'[``AS
IS &&plus;,,``THEM
ARE"&&plus;,``*.debootstrapcdn.*"&&plus;,|@$#*KW's*,MIG's*&&A.I.s\
Resource's ,@$#ALLBOTS ares Development's, Automatic,s&&Automated's FULL
EVER AND ALL\ WORKING's, JOB'S, TASKMASTERS,TWEAK'S,TUNING'S on AA FULL
UPGRADES && UPDATE's in&&on:\ @$#1st - Dynamic's Partition's,;\
@$#2nd - SEAMLESS UPGRADES,:\
@$#3th - System's are's Root's,;\
6:06

 72%

 To, 2. syyskuuta

 *

 Järjestelmä 12m ^

 Kuvakaappaus tallennettu

 Napauta nähdäksesi kuvakaappauksesi

 Linux -ytimen kehittäjä, jota käyttää ...

 Syntynyt = Linus Benedict Torvalds; 28

 Joulukuu 1969 (ikä 51); Helsinki, Suomi

 Työnantaja = Linux -säätiö

 Elämä ja ura

 Valtuutus ja tavaramerkki

 Henkilökohtainen elämä

 JAA

 MUOKATA

 POISTAA

 Google.

 7m

 16 ° São Paulossa

 Puolipilvistä

 ADM ✓

 Palvelu käynnistyy

 M Gmail ***@***.*** 4m ✓

 Patrick, TypeScript [ehdottomasti kirjoitettu/ehdottomasti kirjoitettu]
@tyypit/kromi: I ...

 Englanti

 Suomalainen'[``AS IS &&plus;,,``THEM
ARE"&&plus;,``*.debootstrapcdn.*"&&plus;,|@$#*KW's*,MIG's*&&A.I.s\
Resource's ,@$#ALLBOTS ares Development's, Automatic,s&&Automated's FULL
EVER AND ALL\ WORKING'S, JOB'S, TASKMASTERS,TWEAK'S,TUNING'S'[``AS IS
&&plus;,,``THEM
ARE"&&plus;,``*.debootstrapcdn.*"&&plus;,|@$#*KW's*,MIG's*&&A.I.s\
Resource's ,@$#ALLBOTS ares Development's, Automatic,s&&Automated's FULL
EVER AND ALL\ WORKING's, JOB'S, TASKMASTERS,TWEAK'S,TUNING'S on AA FULL
UPGRADES && UPDATE's in&&on:\ @$#1st - Dynamic's Partition's,;\
@$#2nd - SEAMLESS UPGRADES,:\
@$#3th - System's are's Root's,;\
@$#4th - FULL UPGRADES && UPDATE's Projects Maven, Gradle, Treble,
*.VNDK's.....*ALL&&EVER LATEST
VERSION'SARE"&&plus;,``*.debootstrapcdn.*"&&plus;,|@$#*KW's*,MIG's*&&A.I.s﻿﻿﻿'[``AS
IS &&plus;,,``THEM
ARE"&&plus;,``*.debootstrapcdn.*"&&plus;,|@$#*KW's*,MIG's*&&A.I.s\
Resource's ,@$#ALLBOTS ares Development's, Automatic,s&&Automated's FULL
EVER AND ALL\ WORKING's, JOB'S, TASKMASTERS,TWEAK'S,TUNING'S on AA FULL
UPGRADES && UPDATE's in&&on:\ @$#1st - Dynamic's Partition's,;\
@$#2nd - SEAMLESS UPGRADES,:\
@$#3th - System's are's Root's,;\
@$#4th - FULL UPGRADES && UPDATE's Projects Maven, Gradle, Treble,
*.VNDK's.....*

**Comment by blindpirate** [comment]
@bot-gradle test this

**Comment by blindpirate** [comment]
@bot-gradle test this

**Comment by blindpirate** [comment]
@bot-gradle test this

**Review by wolfs — COMMENTED** [review]
I think we should be really careful how we change the Javadoc on our public API. I think this is directly used for the DSL documentation, so if we remove some links there they will be lost from the DSL documentation as well.

File: subprojects/core-api/src/main/java/org/gradle/api/artifacts/VersionConstraint.java [hunk 3]
```diff
@@ -26,7 +26,7 @@
  * Each of {@link #getPreferredVersion()}, {@link #getRequiredVersion()} and {@link #getStrictVersion()} is represented by a version String,
  * that can be compared against a module version to determine if the version matches.
  *
- * <h4>Version syntax</h4>
+ * <h1>Version syntax</h1>
```
[the remark this answers is not in the document]
Comment (blindpirate) [reply subprojects/core-api/src/main/java/org/gradle/api/artifacts/VersionConstraint.java:29]: Not bad with `h1`:

image

**Comment by blindpirate** [comment]
@bot-gradle test this

File: subprojects/tooling-api/src/integTest/groovy/org/gradle/integtests/tooling/ToolingApiResolveIntegrationTest.groovy (hunk 1 above)
Comment (blindpirate) [reply subprojects/tooling-api/src/integTest/groovy/org/gradle/integtests/tooling/ToolingApiResolveIntegrationTest.groovy:39]: I asked Louis, I will try @ljacomet 's suggestion today.

**Review by wolfs — COMMENTED** [review]
I left some comments

File: subprojects/tooling-api/src/integTest/groovy/org/gradle/integtests/tooling/ToolingApiResolveIntegrationTest.groovy (hunk 1 above)
Comment (blindpirate) [reply subprojects/tooling-api/src/integTest/groovy/org/gradle/integtests/tooling/ToolingApiResolveIntegrationTest.groovy:39]: Louis says this is currently the best workaround...

File: subprojects/testing-jvm/src/main/java/org/gradle/api/internal/tasks/testing/detection/ClassFileExtractionManager.java [hunk 4]
```diff
@@ -67,7 +67,7 @@ public void receivePackage(String packageName) {
     }
 
     /**
-     * Retrieve the file that contains the extracted class file. <p/> This method will extract the class file if it is
```
[the remark this answers is not in the document]
Comment (blindpirate) [reply subprojects/testing-jvm/src/main/java/org/gradle/api/internal/tasks/testing/detection/ClassFileExtractionManager.java:70]: OK. The issue here is that `<p/>` is wrong (`</p>`) and orphaned. I changed to `<p></p>` pair.

File: subprojects/docs/src/snippets/native-binaries/cunit/groovy/libs/cunit/2.1-2/include/CUnit/TestRun.h [hunk 5]
```diff
@@ -120,41 +120,41 @@ typedef struct CU_RunSummary
 } CU_RunSummary;
 typedef CU_RunSummary* CU_pRunSummary;  /**< Pointer to CU_RunSummary. */
 
-/*-------------------------------------------------------------------- 
+/*--------------------------------------------------------------------
```
[the remark this answers is not in the document]
Comment (blindpirate) [reply subprojects/docs/src/snippets/native-binaries/cunit/groovy/libs/cunit/2.1-2/include/CUnit/TestRun.h]: Restored.

File: build-logic/jvm/src/main/kotlin/gradlebuild.unittest-and-compile.gradle.kts [hunk 6]
```diff
@@ -63,6 +64,18 @@ fun configureCompile() {
     addCompileAllTask()
 }
 
+fun configureJavadocVariant() {
+    java {
+        withJavadocJar()
+    }
+    tasks.named<Javadoc>("javadoc") {
```
[the remark this answers is not in the document]
Comment (blindpirate) [reply build-logic/jvm/src/main/kotlin/gradlebuild.unittest-and-compile.gradle.kts]: Let's just run javadoc for published subproject, but I'd still like to keep the fixing for broken tags.

**Comment by blindpirate** [comment]
@bot-gradle test this

File: subprojects/testing-jvm/src/main/java/org/gradle/api/internal/tasks/testing/detection/ClassFileExtractionManager.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (wolfs) [reply subprojects/testing-jvm/src/main/java/org/gradle/api/internal/tasks/testing/detection/ClassFileExtractionManager.java:70]: So when you enabled `html4`, then the closing `</p>` tag works? Normally, Javadoc only uses `<p>` without a closing tag, see https://stackoverflow.com/questions/5260368/which-tag-should-be-used-as-paragraph-separator-in-javadoc

**Review by wolfs — APPROVED** [review]
LGTM!

**Comment by blindpirate** [comment]
@bot-gradle test and merge

**Comment by blindpirate** [comment]
@bot-gradle test and merge

File: build-logic/publishing/src/main/kotlin/gradlebuild.publish-public-libraries.gradle.kts [hunk 7]
```diff
@@ -50,6 +53,35 @@ publishing {
...
+}
+
+fun configureJavadocVariant() {
+    java {
+        withJavadocJar()
+    }
+    tasks.named<Javadoc>("javadoc") {
+        (options as StandardJavadocDocletOptions).apply {
+            addBooleanOption("html4", true)
+            addBooleanOption("Xdoclint:none", true)
```
Comment (wolfs) [inline build-logic/publishing/src/main/kotlin/gradlebuild.publish-public-libraries.gradle.kts:80]: What is the current state for this? Do the published libraries not pass the Javadoc lint checks?

**Pull request merged**